.emacs.d/config.org

```diff
@@ -55,62 +55,392 @@
+    :custom
+    (lsp-auto-guess-root t)
+    (lsp-log-io t)
+    (lsp-prefer-flymake nil))
+  #+END_SRC
+
+  This package enhances the user interface, providing pop-up windows with documentation and visual indicators.
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package lsp-ui
+    :hook (lsp-mode . lsp-ui-mode)
+    :custom
+    (lsp-ui-doc-delay 3.0)
+    (lsp-ui-doc-include-signature t)
+    (lsp-ui-flycheck-enable t)
+    (lsp-ui-peek-always-show t))
+  #+END_SRC
+
+  Analogous to LSP is the [[https://microsoft.github.io/debug-adapter-protocol/][debug adapter protocol]]. We use [[https://github.com/emacs-lsp/lsp-treemacs][treemacs]] for IDE-like display of errors.
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package dap-mode
+    :after lsp-mode
+    :config
+    (dap-mode t)
+    (dap-ui-mode t))
+
+  (use-package treemacs
+    :bind (:map global-map
+                ("C-x t t" . treemacs)
+                ("C-x t 1" . treemacs-select-window))
+    :custom
+    (treemacs-resize-icons 15))
+
+  (use-package lsp-treemacs
+    :init (lsp-treemacs-sync-mode 1)
+
+    :bind (:map java-mode-map
+                ("C-x e l" . lsp-treemacs-errors-list)
+                ("C-x s l" . lsp-treemacs-symbols)))
+  #+END_SRC
+
+* Code completion
+
+  Company is the primary package that is used for code completion, it follows a frontend/backend system. The package =company= is the frontend, it will query a certain backend based on what code you are editing, such as one provided by an active language server.
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package company
+    :hook (after-init . global-company-mode)
+    :custom
+    (company-tooltip-align-annotations t))
+
+  (use-package company-lsp
+    :after (company lsp-mode)
+    :custom
+    (company-lsp-cache-candidates t))
+  #+END_SRC
+
+* Languages
+
+** HTML / CSS / HTML Templates
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package emmet-mode
+    :hook ((css-mode php-mode sgml-mode rjsx-mode web-mode) . emmet-mode))
+  #+END_SRC
+
+** JavaScript / TypeScript
+
+  These two are probably the hardest configuration to get properly solely because of "standards".
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package typescript-mode
+    :hook
+    (typescript-mode . lsp)
+    :mode (("\\.ts\\'" . typescript-mode)
+           ("\\.tsx\\'" . typescript-mode)))
+  #+END_SRC
+
+  Use Node modules if present. Especially helpful for versions of tools and for stuff like prettier.
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package add-node-modules-path
+    :hook ((web-mode rjsx-mode). add-node-modules-path))
+  #+END_SRC
+
+  Actually add prettier (which will read from =.prettierrc= if it exists now). Note you’ll have to install it using =npm install -g prettier=.
+
+  #+BEGIN_SRC emacs-lisp
+  (use-package prettier-js
+    :hook ((js-mode typescript-mode rjsx-mode) . prettier-js-mode))
+  #+END_SRC
+
+  [[https://github.com/ananthakumaran/tide][TypeScript Interactive Development Environment for Emacs]]
+
+  #+BEGIN_SRC emacs-lisp
+  ;; TODO clean this up?
```

Comment:
Nope, it's from official documents mostly

Reply:
Just because something is official doesn't mean we can't make it better. The reason I put the comment there is because *my* solution would be
```lisp
(use-package flycheck
  :hook (after-init . global-flycheck-mode)
  :custom
  (flycheck-check-syntax-automatically '(save mode-enabled)))

(use-package eldoc-mode
  :hook (after-init . global-eldoc-mode))

(use-package company-mode
  :hook (after-init . global-company-mode))
```
and then the hook for `tide-hl-identifier-mode` is already present in the `tide` block. Most people probably want company, flycheck, and eldoc enabled globally.